src/v2/guide/composition.md

```diff
@@ -0,0 +1,546 @@
+      <span slot="bold">Error!</span>
+      <slot/>
+    </AppNote>
+  `
+}
+```
+
+Instead of inheriting and building on the interface of a base component, this strategy does **not** provide direct access to the wrapped component. So unlike the `ErrorNote` using extension, this one has its own props (currently none) and its own slots (currently just the default slot).
+
+This is useful when we'd really like to _remove_ options from a base interface. For example, in the case of `ErrorNote`, it'd be very strange for it to be used like this:
+
+``` html
+<error-note type="success">
+  <span slot="bold">Welcome!</span>
+  Would you like to take the tour?
+</error-note>
+```
+
+Above, a developer has hijacked the component for a new kind of note that is not actually an error. The wrapper version of this component more strictly controls the API, so does not allow this kind of abuse.
+
+<p class="tip">In the example above, <code>type="success"</code> <em>would</em> actually affect the wrapped <code>AppNote</code> component, but as an attribute rather than a prop. To prevent attributes from automatically being added to the root element of a component, use the <a href="../api/#inheritAttrs"><code>inheritAttrs: false</code></a> option.</p>
+
+
+
+## Factory Functions
+
+**Factory functions** are functions that return component options, like this:
+
+``` js
+function createComponent (options) {
+  return {
+    // ... dynamically generated component config ...
+  }
+}
+```
+
+Factory functions allow you to better control and simplify how component variations are created. For example, it could be argued that the `ErrorNote` we [created earlier with extension](#An-Example-Use-Case) would be conceptually simpler as a factory function:
+
+``` js
+function createNote (options) {
+  options = options || {}
+  options.props = options.props || {}
+  options.slots = options.slots || {}
+
+  return {
+    props: {
+      type: {
+        type: String,
+        default: options.props.type || 'info'
+      }
+    },
+    template: `
+      <p :class="['note', type]">
+        <strong>
+          <slot name="bold">${
+            options.slots.bold || ''
+          }</slot>
+        </strong>
+        <slot>${
+          options.slots.default || ''
+        }</slot>
+      </p>
+    `
+  }
+}
+```
+
+Then we could create our `AppNote` and `ErrorNote` components with:
+
+``` js
+var AppNote = createNote()
+
+var ErrorNote = createNote({
+  props: {
+    type: 'error'
+  },
+  slots: {
+    bold: 'Error!'
+  }
+})
+```
+
+Using this strategy, all notes in the app maintain unified behavior, but _you_ control the API for creating them. This can be especially useful when you want to:
+
+- Create many variations of a component, abstracting away the specific implementation details.
+- Give designers, less experienced developers, or members of a very large team the ability to create new component variations, but strictly control what they're allowed to change.
```

Comment:
How would this strictly control what they're allowed to change?  They could change the returned object, right?

Reply:
It restricts the changes that can be made to component options through the "normal" means (i.e. the API exposed by the function's arguments). Changing the object generated by a factory function built to abstract away the concern would be considered not normal and easier to clearly identify as a smell.

There are additional measures one could take if you're worried about this, such as freezing the returned object - but really, nothing can keep you safe from a colleague that doesn't respect exposed APIs. 😅 They could just as easily overwrite built-in browser methods or delete internal Vue properties.